internal/k8s/ingressstatus.go

```diff
@@ -14,45 +14,94 @@
 package k8s
 
 import (
+	"github.com/projectcontour/contour/internal/annotation"
 	"github.com/sirupsen/logrus"
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/api/networking/v1beta1"
 	clientset "k8s.io/client-go/kubernetes"
 )
 
-// StatusLoadbalancerUpdater observes informer OnAdd events and
+// IngressStatusUpdater observes informer OnAdd events and
 // updates the ingress.status.loadBalancer field on all Ingress
 // objects that match the ingress class (if used).
 type IngressStatusUpdater struct {
-	Client clientset.Interface
-	Logger logrus.FieldLogger
-	Status v1.LoadBalancerStatus
+	Client       clientset.Interface
+	Logger       logrus.FieldLogger
+	Status       v1.LoadBalancerStatus
+	IngressClass string
 }
 
 func (s *IngressStatusUpdater) OnAdd(obj interface{}) {
-	ing := obj.(*v1beta1.Ingress).DeepCopy()
 
-	// TODO(dfc) check ingress class
+	ing := obj.(*v1beta1.Ingress).DeepCopy()
+	if !annotation.MatchesIngressClass(ing, s.IngressClass) {
+		s.Logger.
+			WithField("name", ing.GetName()).
+			WithField("namespace", ing.GetNamespace()).
+			WithField("ingress-class", annotation.IngressClass(ing)).
+			Debug("Unmatched ingress class, skip status update")
+		return
+	}
 
 	ing.Status.LoadBalancer = s.Status
 	_, err := s.Client.NetworkingV1beta1().Ingresses(ing.GetNamespace()).UpdateStatus(ing)
 	if err != nil {
 		s.Logger.
 			WithField("name", ing.GetName()).
 			WithField("namespace", ing.GetNamespace()).
-			WithError(err).Error("unable to update status")
+			WithError(err).Error("Unable to update status")
 	}
 }
 
 func (s *IngressStatusUpdater) OnUpdate(oldObj, newObj interface{}) {
-	// Ignoring OnUpdate allows us to avoid the message generated
-	// from the status update.
-
-	// TODO(dfc) handle these cases:
-	// - OnUpdate transitions from an ingress class which is out of scope
-	// to one in scope.
-	// - OnUpdate transitions from an ingress class in scope to one out
-	// of scope.
+
+	oldIng := oldObj.(*v1beta1.Ingress).DeepCopy()
+	newIng := newObj.(*v1beta1.Ingress).DeepCopy()
+
+	// It's important that we only act when we are sure that there has been a change in the
+	// annotation between the old and new versions, otherwise we could generate a status update
+	// in response to another status update.
+	if !annotation.MatchesIngressClass(newIng, s.IngressClass) {
```

Comment:
I think it would be clearer to explicitly spell our the two case that we care about here:

```
if matches(old) && !matches(new) {
    // Clear status because we stopped matching.
}

if !matches(old) && matches(new) {
    // Add status because we started matching.
}
```

I think that this is clearer because the checks are always ordered old -> new, and both conditions are fully specified, so you don't have to think about the previous one when looking at the second check.

Reply:
Yeah, okay, that makes sense.